src/mono/mono/mini/interp/transform.c

```diff
@@ -8364,49 +8373,130 @@ interp_mark_reachable_bblocks (TransformData *td)
 	}
 }
 
+/**
+ * Returns TRUE if in_bb defines a variable that is used in
+ * the block containing @cond_ins, FALSE otherwise.
+ */
+
 static gboolean
-interp_prev_ins_defines_var (InterpInst *ins, int var1, int var2)
+interp_prev_block_defines_var (InterpInst *in_bb_ins, InterpInst *cond_ins)
 {
+	InterpInst *prev_ins = interp_prev_ins(cond_ins);
+	int var1 = -1, var2 = -1;
+	if (MINT_IS_BINOP_CONDITIONAL_BRANCH(cond_ins->opcode)) {
+		var1 = cond_ins->sregs [0];
+		var2 = cond_ins->sregs [1];
+	} else {
+		if (prev_ins && MINT_IS_COMPARE(prev_ins->opcode)) {
+			var1 = prev_ins->sregs [0];
+			var2 = prev_ins->sregs [1];
+		} else {
+			var1 = cond_ins->sregs [0];
+		}
+	}
+
 	// Check max of 5 instructions
+	prev_ins = in_bb_ins;
 	for (int i = 0; i < 5; i++) {
-		ins = interp_prev_ins (ins);
-		if (!ins)
+		prev_ins = interp_prev_ins (prev_ins);
+		if (!prev_ins)
 			return FALSE;
-		if (mono_interp_op_dregs [ins->opcode] && (ins->dreg == var1 || ins->dreg == var2))
+		if (mono_interp_op_dregs [prev_ins->opcode] && (prev_ins->dreg == var1 || prev_ins->dreg == var2))
 			return TRUE;
 	}
 	return FALSE;
 }
 
+/**
+ * Check if the given basic block has a known pattern for inlining into callers blocks, if so, return a pointer to the conditional branch instruction.
+ *
+ * The known patterns are:
+ * - `branch`: a conditional branch instruction.
+ * - `ldc; branch`: a load instruction followed by a binary conditional branch.
+ * - `compare; branch`: a compare instruction followed by a unary conditional branch.
+ * - `ldc; compare; branch`: a load instruction followed by a compare instruction and a unary conditional branch.
+ */
+static InterpInst*
+interp_inline_into_callers(InterpInst *first) {
+	// pattern `branch`
+	if (MINT_IS_CONDITIONAL_BRANCH(first->opcode)) {
+		return first;
+	}
+	
+	InterpInst *second = interp_next_ins(first);
+	if (!second) 
+		return NULL;
+	// pattern `ldc; unop conditional branch`
+	if (MINT_IS_LDC(first->opcode)
```

Comment:
This pattern is redundant and it is probably not reached anyway. `ldc` + `brtrue/false` is already resolved to `br.s` or optimized out.

Reply:
Good point. Removed.